frontend/server/handlers/artifacts.ts

```diff
@@ -237,3 +238,89 @@ function getGCSArtifactHandler(options: { key: string; bucket: string }, peek: n
     }
   };
 }
+
+const AUTH_EMAIL_HEADER = 'x-goog-authenticated-user-email';
+const ARTIFACTS_PROXY_DEFAULTS = {
+  serviceName: 'ml-pipeline-ui-artifact',
+  servicePort: '80',
+};
+export interface ArtifactsProxyConfig {
+  serviceName: string;
+  servicePort: number;
+  enabled: boolean;
+}
+export function loadArtifactsProxyConfig(env: ProcessEnv): ArtifactsProxyConfig {
+  const {
+    ARTIFACTS_SERVICE_PROXY_NAME = ARTIFACTS_PROXY_DEFAULTS.serviceName,
+    ARTIFACTS_SERVICE_PROXY_PORT = ARTIFACTS_PROXY_DEFAULTS.servicePort,
+    ARTIFACTS_SERVICE_PROXY_ENABLED = 'false',
+  } = env;
+  return {
+    serviceName: ARTIFACTS_SERVICE_PROXY_NAME,
+    servicePort: parseInt(ARTIFACTS_SERVICE_PROXY_PORT, 10),
+    enabled: ARTIFACTS_SERVICE_PROXY_ENABLED.toLowerCase() === 'true',
+  };
+}
+
+const QUERIES = {
+  NAMESPACE: 'namespace',
+};
+
+export function getArtifactsProxyHandler({
+  serviceName,
+  servicePort,
+  enabled,
+}: ArtifactsProxyConfig): Handler {
+  if (!enabled) {
+    return (req, res, next) => next();
+  }
+  return proxy(
+    (_pathname, req) => {
+      // only proxy requests with namespace query parameter
+      return !!getNamespaceFromUrl(req.url || '');
+    },
+    {
+      changeOrigin: true,
+      onProxyReq: proxyReq => {
+        console.log('Proxied artifact request: ', proxyReq.path);
+      },
+      pathRewrite: (pathStr, req) => {
+        const url = new URL(pathStr || '', DUMMY_BASE_PATH);
+        url.searchParams.delete(QUERIES.NAMESPACE);
+        return url.pathname + url.search;
+      },
+      router: req => {
+        const namespace = getNamespaceFromUrl(req.url || '');
+        if (!namespace) {
+          throw new Error(`namespace query param expected in ${req.url}.`);
+        }
+        // HACK: tests can spyOn TEST_ONLY to replace implementation.
+        return TEST_ONLY.getArtifactFetcherService({ serviceName, servicePort }, namespace);
```

Comment:
nit: I'm not familiar with frontend to suggest any alternative to avoid this. It is really hacky and usually better to avoid having the test back door in production code.

Reply:
Totally agree, let me think about how to avoid this. We'd probably need some simple dependency injection to open the door for testing.

Reply:
@chensun I refactored and passed in a function instead. I think that's better.

Reply:
yep, looking good. thanks!